django/db/models/deletion.py

```diff
@@ -219,7 +220,30 @@ def collect(self, objs, source=None, nullable=False, collect_related=True,
                     if self.can_fast_delete(sub_objs, from_field=field):
                         self.fast_deletes.append(sub_objs)
                     elif sub_objs:
-                        field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        try:
+                            field.remote_field.on_delete(self, field, sub_objs, self.using)
+                        except ProtectedError as e:
+                            protected_objects.append({'field': field, 'sub_objs': sub_objs})
+            number_of_protected_objects = len(protected_objects)
+            if number_of_protected_objects == 1:
```

Comment:
Try to avoid two branches and lots of code duplication here. You should be able to handle both cases in single path.